### PR TITLE
Refactor: Improve Event Tickets API validations

### PR DESCRIPTION
**Pull request by pauloiankoski**
Resolves [GIVE-371]

## Description

This pull request improves Event Tickets API routes by enhancing permission checks and validations to prevent invalid requests from reaching the handler. Additionally, it includes unit tests that cover various edge cases, that ran through the actual API system with the `RestApiTestCase`.

## Affects

Event Tickets API routes

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-371]: https://stellarwp.atlassian.net/browse/GIVE-371?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/EventTickets/Routes/DeleteEventTicketType.php [hunk 1]
```diff
@@ -49,36 +55,25 @@ public function registerRoute()
...
     {
         $ticketType = EventTicketType::find($request->get_param('ticket_type_id'));
 
         $salesCount = $ticketType->eventTickets()->count();
 
         if ($salesCount > 0) {
-            return new WP_REST_Response([
-                'message' => __('This ticket type has been sold and cannot be deleted.', 'give'),
-            ], 400);
+            return new WP_Error(
```
Comment (kjohnson) [inline src/EventTickets/Routes/DeleteEventTicketType.php:68]: What is the benefit in changing this from a `WP_REST_Response` to a `WP_Error`?

File: src/EventTickets/Routes/GetEventForms.php [hunk 2]
```diff
@@ -28,7 +31,9 @@ public function registerRoute()
                 [
                     'methods' => 'GET',
                     'callback' => [$this, 'handleRequest'],
-                    'permission_callback' => '__return_true',
+                    'permission_callback' => function () {
+                        return current_user_can('read');
```
Comment (kjohnson) [inline src/EventTickets/Routes/GetEventForms.php:35]: I would think that this endpoint is either public (no permissions needed), or only readable by an administrator. Checking against the `read` capability suggests that the user must be authenticated, but any authenticated user can access the information, which doesn't sound correct.

File: src/EventTickets/Routes/DeleteEventTicketType.php [hunk 3]
```diff
@@ -30,14 +34,16 @@ public function registerRoute()
...
+                    },
                 ],
                 'args' => [
                     'ticket_type_id' => [
                         'type' => 'integer',
                         'sanitize_callback' => 'absint',
-                        'validate_callback' => function ($eventId) {
-                            return EventTicketType::find($eventId);
+                        'validate_callback' => function ($ticketTypeId) {
+                            return give(EventTicketTypeRepository::class)->queryById($ticketTypeId)->count() > 0;
```
Comment (kjohnson) [inline src/EventTickets/Routes/DeleteEventTicketType.php:46]: What is the benefit of validating against the count, as opposed to `EventTicketType|null`? Using the facade, if the entity does not exist then it returns null, which should fail the validation.

Also, if we are to validate against the count, comparing against zero is not necessary, since a positive integer is already `truthy` and zero is `falsey`.

File: src/EventTickets/Routes/CreateEvent.php [hunk 4]
```diff
@@ -25,10 +29,10 @@ public function registerRoute()
             $this->endpoint,
             [
                 [
-                    'methods' => 'POST',
+                    'methods' => WP_REST_Server::CREATABLE,
                     'callback' => [$this, 'handleRequest'],
                     'permission_callback' => function () {
-                        return current_user_can( 'manage_options' );
+                        return current_user_can('publish_give_payments');
```
Comment (kjohnson) [inline src/EventTickets/Routes/CreateEvent.php:35]: Since we don't have a specific permission for this dataset, does it make sense to authorize against a payment capability? Does it suffice to fallback to a simple "is admin" (`manage_options`) check?

File: src/EventTickets/Routes/GetEventForms.php (hunk 2 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/GetEventForms.php:35]: That's a valid point. I suggest restricting this to admins who can manage forms and posts. What level of capability do you think is appropriate for this? Should we limit it to `manage_options` or consider a lower level like `edit_posts`?

File: src/EventTickets/Routes/CreateEvent.php (hunk 4 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/CreateEvent.php:35]: I think the answer in the second comment applies to this as well.

File: src/EventTickets/Routes/DeleteEventTicketType.php (hunk 3 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/DeleteEventTicketType.php:46]: I only noticed the issue when writing unit tests. The validation callback was returning `null` for invalid ticket type ids without halting execution. The reason is that the `validate_callback` checks for a false return from that method, and `null` is not considered false.

CleanShot 2024-03-26 at 17 36 03

A solution would be check against `null` like in the snippet bellow. What do you think?
```php
'validate_callback' => function ($ticketTypeId) {
    return EventTicketType::find($ticketTypeId) !== null;
},

Comment (kjohnson) [reply src/EventTickets/Routes/DeleteEventTicketType.php:46]: Yeah, I think it would be simpler to check against `null` and leverage our facade.

As an aside, we might consider a `::findOrFail($id)` method in the future (another good Friday project).

File: src/EventTickets/Routes/GetEventForms.php (hunk 2 above)
Comment (kjohnson) [reply src/EventTickets/Routes/GetEventForms.php:35]: Using `manage_options` is a common fallback for "is administrator", which I usually prefer over re-using `edit_posts` (when it isn't a `post`).

File: src/EventTickets/Routes/DeleteEventTicketType.php (hunk 1 above)
Comment (kjohnson) [reply src/EventTickets/Routes/DeleteEventTicketType.php:68]: Interesting point. In that case, should this check be moved to the attribute `validate_callback`?

File: src/EventTickets/Routes/GetEventForms.php (hunk 2 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/GetEventForms.php:35]: I'm concerned that users with the `give_manager` role can't manage events using `manage_options`. I'm exploring a capability like `manage_give_settings` that could meet this need for both Administrators and Give Managers.

Comment (kjohnson) [reply src/EventTickets/Routes/GetEventForms.php:35]: Ok, if you find a better capability then let's use that.

File: src/EventTickets/Routes/DeleteEventTicketType.php (hunk 3 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/DeleteEventTicketType.php]: Changed all to be like `Model::find($entityId) !== null` on ef40760a21e94d8b887460706f6b97d63be4915a

File: src/EventTickets/Routes/DeleteEventTicketType.php (hunk 1 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/DeleteEventTicketType.php]: I do agree. Moved it to the `validation_callback` while keeping the custom message.

File: src/EventTickets/Routes/GetEventForms.php (hunk 2 above)
Comment (pauloiankoski) [reply src/EventTickets/Routes/GetEventForms.php]: After checking that other menu items use the `edit_give_forms` to even being displayed, the permission checks were updated to use that capability: ef40760a21e94d8b887460706f6b97d63be4915a

**Review by kjohnson — APPROVED** [review]
